go-controller/pkg/util/multinetwork.go

```diff
@@ -0,0 +1,262 @@
+// Note that for port_group and address_set, it does not allow the '-' character,
+// which will be replaced with ".". Also replace "/" in the nadName with "."
+func GetSecondaryNetworkPrefix(netName string) string {
+	name := strings.ReplaceAll(netName, "-", ".")
+	name = strings.ReplaceAll(name, "/", ".")
+	return name + "_"
+}
+
+func newNetConfInfo(netconf *ovncnitypes.NetConf) (NetConfInfo, error) {
+	var netconfInfo NetConfInfo
+
+	if !netconf.IsSecondary {
+		return nil, nil
+	}
+
+	if netconf.Topology == types.Layer3AttachDefTopoType {
+		netconfInfo = &Layer3NetConfInfo{
+			NetCidr: netconf.NetCidr,
+			MTU:     netconf.MTU,
+		}
+	} else {
+		// other topology nad can be supported later
+		return nil, fmt.Errorf("topology %s not supported", netconf.Topology)
+	}
+	err := netconfInfo.Verify()
+	if err != nil {
+		return nil, fmt.Errorf("%s netconf verification failed: %v", netconf.Topology, err)
+	}
+	return netconfInfo, nil
+}
+
+// ParseNADInfo parses config in NAD spec and return a NetAttachDefInfo object for secondary networks
+func ParseNADInfo(netattachdef *nettypes.NetworkAttachmentDefinition) (NetInfo, NetConfInfo, error) {
+	netconf, err := ParseNetConf(netattachdef)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	netconfInfo, err := newNetConfInfo(netconf)
+	if err != nil {
+		return nil, nil, err
+	}
+	return NewNetInfo(netconf), netconfInfo, nil
+}
+
+func NewNetInfo(netconf *ovncnitypes.NetConf) NetInfo {
+	// default network netInfo is nil
+	nInfo := (*NetNameInfo)(nil)
+	if netconf.IsSecondary {
+		nInfo = &NetNameInfo{
+			netName:  netconf.Name,
+			nadNames: &sync.Map{},
+		}
+	}
+	return nInfo
+}
+
+// ParseNetConf parses config in NAD spec for secondary networks
+func ParseNetConf(netattachdef *nettypes.NetworkAttachmentDefinition) (*ovncnitypes.NetConf, error) {
+	netconf, err := config.ParseNetConf([]byte(netattachdef.Spec.Config))
+	if err != nil {
+		return nil, fmt.Errorf("error parsing Network Attachment Definition %s/%s: %v", netattachdef.Namespace, netattachdef.Name, err)
+	}
+	// skip non-OVN nad or primary network OVN NAD
+	if netconf.Type != "ovn-k8s-cni-overlay" || !netconf.IsSecondary {
+		return nil, ErrorAttachDefNotOvnManaged
+	}
+
+	nadName := GetNadName(netattachdef.Namespace, netattachdef.Name)
+	if netconf.NadName != nadName {
+		return nil, fmt.Errorf("net-attach-def name (%s) is inconsistent with config (%s)", nadName, netconf.NadName)
+	}
+
+	return netconf, nil
+}
+
+// IsNetworkOnPod sees if the given pod needs to plumb over this given network specified by netconf,
+// and return the matching NetworkSelectionElement if any exists.
+//
+// Return value:
+//    bool: if this Pod is on this Network; true or false
+//    *networkattachmentdefinitionapi.NetworkSelectionElement: all NetworkSelectionElement that pod is requested for the specified network
+//    error:  error in case of failure
+func IsNetworkOnPod(pod *kapi.Pod, nInfo NetInfo) (bool, *nettypes.NetworkSelectionElement, error) {
```

Comment:
How do we handle updates to the network selection annotation or default network annotation. I haven't seen this covered in the spec.
For example, what happens if through a pod update, a network that was previously annotated on the pod is removed?

Reply:
that is not supported. does multus handles that?

Reply:
No, multus only cares about the result of the attachment operation (CNI ADD) and removal (CNI DEL).

I've worked in an extension for multus to allow network hotplug for pods in this repo; it still does **not** handle an update to an existing attachment (only handles requesting new attachments / deleting old attachments).

Reply:
So by being not supported, I guess that it is expected that something breaks if such a thing happens right?

@cathy-zhou any reason to not have the web-hook prevent such changes, just as you have one preventing removal of NADs that are still in use?

Reply:
> So by being not supported, I guess that it is expected that something breaks if such a thing happens right?

No. Given the current architecture, **nothing** will happen - i.e. there's nothing reacting to that update - technologically speaking - and the behavior is not defined in the spec. @dougbtv care to keep me honest ? 

> 
> @cathy-zhou any reason to not have the web-hook prevent such changes, just as you have one preventing removal of NADs that are still in use?

Reply:
I guess it is possible to have such web-hook, but in real-life deployment, that is rare and as you said, no harm will be done.

Reply:
But shouldn't we minimally handle this situation? For example, to avoid leaking the allocated IPs and other stuff for that pod network attachment that is consuming master's memory and that will otherwise stay there forever unless it is restarted or the NAD deleted? This could be potentially harmful I guess.

Reply:
but as multus (cni) does not do anything about it, so ip is still used in the pod. also there are other resources (for example, VFs) that is not released by cni.

again, this is not supported behavior, as long as it does not cause harm to the whole system or even this pod, I don't see why we need to anything.

Reply:
Let's say that a tenant is authorized to attach its pods to a secondary network. Then the tenant can:

1. Create a pod attached to this secondary network
2. Update the pod and remove attachment from the networks annotation
3. Delete the pod

Since neither on step 2 or 3 we recognize the pod as being attached to that secondary network, we don't remove the IP allocation, nor the logical port nor any other resource that was basically reserved for that pod. That means that the control plane is leaking resources. If the tenant does that sequence of operations an unlimited number of times, then it can adversely impact the control plane capacity.